### PR TITLE
adding Replay

### PR DESCRIPTION
**Pull request by russellbstephens**

### CODE REVIEW

File: Snail/Observable.swift [hunk 1]
```diff
@@ -3,22 +3,25 @@
 public class Observable<T> : ObservableType {
     public typealias E = T
     private var isStopped: Int32 = 0
-    private var eventHandlers: [(Event<E>) -> Void] = []
+    var eventHandlers: [(Event<E>) -> Void] = []
```
Comment (wesbillman) [inline Snail/Observable.swift]: I don't think we want these to be non-private do we? maybe `internal` (or is that the default)?

File: SnailTests/ReplayTests.swift [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    func testMultipleSubscribers() {
+        var more: [String] = []
+        subject?.on(.next("1"))
+        subject?.on(.next("2"))
+        subject?.on(.next("3"))
+        subject?.subscribe(onNext: { string in
+            more.append(string)
+            self.strings?.append(string)
```
Comment (wesbillman) [inline SnailTests/ReplayTests.swift]: why use both `more` and `strings`?

File: Snail/Replay.swift [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        replay(handler)
+    }
+
+    public override func subscribe(onNext: ((T) -> Void)? = nil, onError: ((Error) -> Void)? = nil, onDone: (() -> Void)? = nil) {
+        super.subscribe(onNext: onNext, onError: onError, onDone: onDone)
+        replay(createHandler(onNext: onNext, onError: onError, onDone: onDone))
+    }
+
+    public override func on(_ event: Event<E>) {
+        events.append(event)
```
Comment (wesbillman) [inline Snail/Replay.swift]: I think we might want to do something like append, then reset the array to be the threshold, so it doesn't just grow forever

```swift
events.append(event)
events = events.suffix(threshold)
```

File: SnailTests/ReplayTests.swift [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        subject = Replay<String>(2)
+        strings = []
+    }
+
+    override func tearDown() {
+        super.tearDown()
+    }
+
+    func testEvent() {
+        var more: [String] = []
```
Comment (wesbillman) [inline SnailTests/ReplayTests.swift]: why more instead of `self.strings`?

File: Snail/Replay.swift [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        guard eventHandlers.count > 0 else {
```
Comment (wesbillman) [inline Snail/Replay.swift]: why do you have this check?

File: Snail/Observable.swift (hunk 1 above)
Comment (russellbstephens) [inline Snail/Observable.swift]: internal is default

File: SnailTests/ReplayTests.swift [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        subject?.on(.next("1"))
+        XCTAssert(strings.first == "1")
+    }
+
+    func testOnNext() {
+        var strings: [String] = []
+        subject?
+            .subscribe(onNext: { string in strings.append(string) })
+        subject?.on(.next("1"))
+        subject?.on(.next("2"))
```
Comment (wesbillman) [inline SnailTests/ReplayTests.swift]: This isn't testing replay. It's just testing normal observables. You need to `on` before you subscribe, right?

File: Snail/Replay.swift (hunk 3 above)
Comment (russellbstephens) [inline Snail/Replay.swift]: we just need to create a new array as well since suffix returns a slice

File: SnailTests/ReplayTests.swift (hunk 6 above)
Comment (russellbstephens) [inline SnailTests/ReplayTests.swift]: i reversed it for some reason!

**Pull request merged**